### ISSUE TITLE
[SUGGESTION] Confusing std::vector ctor calls in Cpp1 and Cpp2

### ISSUE THREAD

**Issue by wolfseifert** [hsutter/cppfront#193]
In Cpp1 we have the confusing fact that
```
#include <vector>
using namespace ::std;
int main() {
  auto a = vector<int>{5, 1}; // vector of five and one
  auto b = vector<int>(5, 1); // vector of five ones
}
```
Experimenting with (pure) Cpp2 I got a similar case
```
main: () -> int = {
  a: std::vector<int> = (5, 1); // vector of five and one
  b := std::vector<int>(5, 1); // vector of five ones
}
```
This becomes clear when looking at the tanspiled code
```
  std::vector<int> a { 5, 1 };  // vector of five and one
  auto b { std::vector<int>(5, 1) }; // vector of five ones
```
For me the Cpp2 case looks even more confusing (than the Cpp1 case), because both ctor calls use parenthesis.

**Comment by MichaelCook** [comment]
Perhaps another way to state this issue is: Should cpp2 do fine-grained removal/repair of unfortunate cpp1 API choices?

**Comment by robertshepherdcpp** [comment]
What about if we had a literal for the size, like `_sv` so then we could just define the vector like:
```C++
a: std::vector<int> = (5, 1_sv);  // a vector of 5 ones
b: std::vector<int> = (5, 1);     // a vector of 5 and 1
```
and then we could do it like this:
```C++
  b := std::vector<int>(5, 1_sv); // vector of five ones
```

**Comment by JohelEGP** [comment]
How would you combine that with existing suffixes like `ULL`?

**Comment by filipsajdak** [comment]
Soon you will be able to write:
```cpp
a: std::vector<int> = (5, 1 as std::size_t);
```

**Comment by robertshepherdcpp** [comment]
@fpelliccioni Yes, thats good. Would an overload for a `std::size_t` at the end of the vector be possible?

**Comment by fpelliccioni** [comment]
> @fpelliccioni Yes, thats good. Would an overload for a `std::size_t` at the end of the vector be possible?

I think you confused me with @filipsajdak

**Comment by JohelEGP** [comment]
What about
```C++
  auto a = vector<std::size_t>{5z, 1z}; // vector of five and one
  auto b = vector<std::size_t>(5z, 1z); // vector of five ones
```

**Comment by filipsajdak** [comment]
@JohelEGP cheater. 🙂

I agree that there is an issue. I am just trying to advertise the cpp2 casting feature.

What we want to achieve is unambiguous syntax. The spotted issue sure will need some reasonable solution.

**Comment by switch-blade-stuff** [comment]
IMO this would be fairly difficult to solve while keeping full compatibility with Cpp1, the issue is that in C++, the aggregate (i.e. `{}`) constructor favors initializer lists, while the "classic" parentheses never generate initializer lists.

So, this would either require a change in constructors of `std::vector` (and other containers too) to outright remove the size & value constructor, or use some kind of tag like with `std::in_place_t` to disambiguate overloads; or to forego the initializer-list/aggregate constructor form in Cpp2, which would prevent the ambiguity, but also disable initializer lists.

**Comment by switch-blade-stuff** [comment]
Honestly, it would be easier if C++ just left `{}` for initializer lists, and only use `()` for constructors. That way there would be no ambiguity, if you saw `std::vector v = {0, 1}`, you'd know it is definitely an initializer list, and the vector contains 0 and 1, instead of 1 zero.

**Comment by marioarbras** [comment]
> Honestly, it would be easier if C++ just left {} for initializer lists, and only use () for constructors. That way there would be no ambiguity, if you saw std::vector v = {0, 1}, you'd know it is definitely an initializer list, and the vector contains 0 and 1, instead of 1 zero.

This would be really nice. An alternative idea could be to use an explicit cast for initializer lists.

```
v: std::vector = (5, 1) as std::initializer_list;
```

**Comment by robertshepherdcpp** [comment]
@marioarbras Or we could have it the other way round, depending on which is more common: initializing a vector with a `std::initializer_list` like:
```C++
  a: std::vector<int> = (5, 1); // vector of five and one
```
or initializing it with a certain amount of elements like:
```C++
  b := std::vector<int>(5, 1); // vector of five ones
```
Which would be the more common usage? Personally I use the initializer list more.

**Comment by hsutter** [comment]
Yes, I have that on my list to address as follows (but it's not implemented yet):

    v1: vector<int> = (5, 5);   // size == 5, same as "= ((5, 5, 5, 5, 5))"
    v2: vector<int> = ((5, 5)); // size == 2

The idea is that the parens in `v2`'s initializer make it explicit that we are constructing with a single argument that is a list.

**Comment by switch-blade-stuff** [comment]
> v: std::vector = (5, 1) as std::initializer_list;

Honestly, to me that looks just ***slightly*** better than manual insertion 😅

I suppose we could just explicitly use braces for initializer lists only, where if `{stuff}` is encountered in a constructor it would be replaced by `({stuff})`, forcing the initializer-list overload.

Then again, this would fall flat for aggregate types such as `std::array`, as well as the "classic" C-style arrays that do not use `std::initializer_list`.

**Comment by switch-blade-stuff** [comment]
> The idea is that the parens in v2's initializer make it explicit that we are constructing with a single argument that is a list.

What about this situation though:

```
v1: vector<int> = (x + 1);
v2: vector<int> = ((x + 1));
```

Is v2 a vector of 1 element x+1, or is it a vector of x+1 zeros, where the size was just enclosed in redundant parentheses?

**Comment by SebastianTroy** [comment]
This could be an opportunity to add a small language feature. In cases where you want multiple parameter packs, it would be nice to have a symbol to disambiguate/delineate where you the caller want one to end and the next to begin, this could also be useful in this case.

I'm not sure what that would look like, but may be worth considering


On 2 January 2023 17:58:46 SwitchBlade ***@***.***> wrote:

IMO this would be fairly difficult to solve while keeping full compatibility with Cpp1, the issue is that in C++, the aggregate (i.e. {}) constructor favors initializer lists, while the "classic" parentheses never generate initializer lists.

So, this would either require a change in constructors of std::vector (and other containers too) to outright remove the size & value constructor, or use some kind of tag like with std::in_place_t to disambiguate overloads; or to forego the initializer-list/aggregate constructor form in Cpp2, which would prevent the ambiguity, but also disable initializer lists.

**Comment by marioarbras** [comment]
> Yes, fixing that problem is on my list to address as follows (I just haven't implemented it yet):
> 
> ```
> v1: vector<int> = (5, 5);   // size == 5, same as "= ((5, 5, 5, 5, 5))"
> v2: vector<int> = ((5, 5)); // size == 2
> ```
> 
> The idea is that the parens in `v2`'s initializer make it explicit that we are constructing with a single argument that is a list. (Do you "see" it visually, that the outer parens now clearly contain only a single thing? I think it's visually clear, but I'm curious to learn whether others will see it that way too.)

This makes a lot of sense since you can already do this:

```
il: auto = (1, 1);
```

which transpires to

```cpp
auto il = { 1, 1 }; // same as auto il = std::initializer_list{ 1, 1 }
```

**Comment by switch-blade-stuff** [comment]
> auto il = { 1, 1 }; // same as auto il = std::initializer_list{ 1, 1 }

Correct me if I'm wrong, but isn't `std::initializer_list` different from just `{}`? Also, to my knowledge initializer lists can't be manually constructed.

**Comment by hsutter** [comment]
@switch-blade-stuff 
> What about this situation though:
> `v1: vector<int> = (x + 1);`
> `v2: vector<int> = ((x + 1));`
> Is v2 a vector of 1 element x+1, or is it a vector of x+1 zeros, where the size was just enclosed in redundant parentheses?

The extra parens would be meaningful, as they're saying that "this argument list contains a single element, which is a list. So:

    v1: vector<int> = (x + 1);   // x+1 elements, all zero
    v2: vector<int> = ((x + 1)); // one element, x+1

What do you think, does that sound reasonable?

**Comment by switch-blade-stuff** [comment]
That does sound reasonable, although to me it still seems like a potentially error-prone thing to use double parentheses for lists, since the majority of people are used to `{}` meaning lists and parentheses meaning precedence groups.

Also gives me a bit of a tingle with templates, I cannot come up with a reasonable example off the top off head right now but it *could* cause some unexpected pitfalls in complex template pack expansion (i.e. pack expansion wants to use parentheses for grouping, but because the pack has only 1 element it gets treated as a list).

**Comment by GPMueller** [comment]
Since functions are declared like mappings between tuples, `(...) -> (...)`, I would find it consistent that assigning such a tuple to a vector, `v: vector<int> = (5, 5)` would produce a vector of two elements.
Not sure whether there's other parts of the language syntax that could be considered in this consistency argument.

Personally, I'd find the extra parens more confusing than the fact that a constructor-call is not necessarily the same as an assignment, so while I agree that the cpp1 example given in the original post has potential for confusion, I don't agree that the cpp2 example does.

I see more risk for confusion with
```cpp
v1: vector<int> = (x + 1);
v2: vector<int> = ((x + 1));
```
since people are probably used to parentheses being only about association/ordering.

**Comment by switch-blade-stuff** [comment]
> since people are probably used to parentheses being only about association/ordering.

I agree. Personally, at least, I have always thought (and felt, maybe?) of parentheses as math-related syntax, with the meaning of expression ordering/grouping. And I would assume that most people think of parentheses as something you use to disambiguate `(a + b) * x` from `a + b * x` first, and as such I would expect that people could find the double parentheses syntax confusing. It wouldnt help that most people would be already used to braces used to define lists.

What also can make it more confusing is the possibility of whitespace in between double parentheses. If `((x))` means the same as `( (x) )` it can create even more confusion than there already is about list initialization. Then on the other hand, if `((x))` and `( (x) )` mean different things, it wouldn't remove any confusion either.

The way I see it is that aggregate/list initialization needs to be dealt with by separating construction from lists, rather than replacing the current status-quo with a different syntax.

**Comment by SebastianTroy** [comment]
@HerbSutter is the plan to reserve curlies {} for scopes only in cpp2?


On 2 January 2023 19:46:27 "G. P. Müller" ***@***.***> wrote:

Since functions are declared like mappings between tuples, (...) -> (...), I would find it consistent that assigning such a tuple to a vector, v: vector<int> = (5, 5) would produce a vector of two elements.
Not sure whether there's other parts of the language syntax that could be considered in this consistency argument.

Personally, I'd find the extra parens more confusing than the fact that a constructor-call is not necessarily the same as an assignment, so while I agree that the cpp1 example given in the original post has potential for confusion, I don't agree that the cpp2 example does.

I see more risk for confusion with

v1: vector<int> = (x + 1);
v2: vector<int> = ((x + 1));

since people are probably used to parentheses being only about association/ordering.

**Comment by HALL9kv0** [comment]
My 3 cents on the issue:
1) Additional parentheses are somewhat confusing, and are easy to forget, neglect, especially for newcommers to the language. 
2) Parentheses construction syntax, v: vector<int> =  (5,1), is similar to a function call, and intuitively we can think of it as a function generating a vector of 5 elements, all equal to 1. Easier to introduce it to new cpp programmers. 
3)  C-style arrays and cpp1 containers (std::vector the most used by far) use {} to initialize their elements. Therefore,  v: vector<int> = {1,2} is easier thought as a list of things I want to construct my vector with.
TL;DR: My personal preference, {} for construction using list of elements, () for everything else. Depending on the bracket style intention is clear.

**Comment by fpelliccioni** [comment]
> Yes, fixing that problem is on my list to address as follows (I just haven't implemented it yet):
> 
> ```
> v1: vector<int> = (5, 1);   // size == 5, same as "= ((1, 1, 1, 1, 1))"
> v2: vector<int> = ((5, 1)); // size == 2
> ```
> 
> The idea is that the parens in `v2`'s initializer make it explicit that we are constructing with a single argument that is a list. (Do you "see" it visually, that the outer parens now clearly contain only a single thing? I think it's visually clear, but I'm curious to learn whether others will see it that way too.)

I think it is still confusing. What about named constructors? 
For containers the default have to be initializer list constructing and named constructors should be used for other kind of constructions.
We need a way to map a Named Constructor in Cpp2 with an existing constructor of current C++.

**Comment by marioarbras** [comment]
@switch-blade-stuff

> > auto il = { 1, 1 }; // same as auto il = std::initializer_list{ 1, 1 }
> 
> Correct me if I'm wrong, but isn't `std::initializer_list` different from just `{}`? Also, to my knowledge initializer lists can't be manually constructed.

I could be wrong, but I think an initializer list is implicitly constructed when you do `auto il = {...}`. I think you can also manually construct them explicitly. Check these examples on [cppreference](https://en.cppreference.com/w/cpp/utility/initializer_list/initializer_list). It looks like both `auto il = { 1, 1 }` and `auto il = std::initializer_list{ 1, 1 }` are allowed.

**Comment by HALL9kv0** [comment]
```
using Vec2D=std::vector<std::vector<int>>;
v1: Vec2D = (( (2,1) ,  ((3,4)) )) // v=[1 1; 3 4]
v2: Vec2D= { (2,1)  , {2,3} } // v=[1 1; 3 4]
v3: Vec2D= cpp2::from( (2,1)  , cpp2::from(2,3) ) // v=[1 1; 3 4]
v4: Vec2D= ( cpp2::vector::construct(2,1)  , (2,3) ) // v=[1 1; 3 4]

```
The above concept code is for illustrative purpose only, to highlight 4 different approaches to the subject. 

1. v1 uses the double-parentheses for construction from elements
2. v2 uses {} for construction from elements and () for the other constructor
3. v3 uses a custom cpp2::from constructor wrapper for the initialization list.
4. v4 uses a custom cpp2::vector::construct constructor wrapper for the size, value constructor, similar to what @fpelliccioni suggests if I understood correctly.

Sorted preference max->min: v2, v3, v4, v1. In nested containers the (()) syntax becomes cumbersome and 4 )))) in a row reminds me of Lisp.
What is your preference?

**Comment by switch-blade-stuff** [comment]
> What is your preference?

I agree with the v2 > v3 > v4 > v1 preference, although to me `cpp2::from` doesnt sound descriptive enough, maybe something like `from_list` or `init_list` instead?

Then again, `{}` is much more descriptive, since `cpp2::from` etc. blends itself with functions and constructors and is harder to notice.

v1 is just way too many parentheses.

**Comment by JohelEGP** [comment]
`std::initializer_list`, the initializer of all evil.

IIRC, in Cpp2, `()` is a parameter list, which is necessary to initialize a `std::vector` via a multi-parameter constructors. That is in contrast to `v : std::string = ""`, which doesn't require `()` around `""`. I suppose `("")` would work, too. So maybe just require `{}` is for `std::initializer_list` parameters only (perhaps not even for aggregates), so `v : std::vector<int> = {5, 1}` is $(5, 1)$ and `v : std::vector<int> = (5, 1)` is $(1, 1, 1, 1, 1)$.

**Comment by switch-blade-stuff** [comment]
> std::initializer_list, the initializer of all evil. 

More like `9.4.4 [dcl.init.list] list-initialization` - the source of all pain :laughing:

**Comment by jcanizales** [comment]
Like others, I think this, while it could be tempting, isn't a good idea:
```c++
v1: vector<int> = (x + 1);   // x+1 elements, all zero
v2: vector<int> = ((x + 1)); // one element, x+1
```
It is one new thing to teach instead of one less thing :no_good: :no_entry_sign: :no_entry: . And it's bound to bite someone in the future that's injecting code on compile time and wants the extra `( )` around an expression they're bringing from somewhere else.

I do favor separating "calling a constructor" from "assigning these aggregated values". Like using only `( )` for the former and only `{ }` for the latter. After all, the syntax for designated initializers is already incompatible with the parentheses syntax.

I also think any newcomer to a language would be unpleasantly surprised if `v: vector = (5,1)` didn't initialize a vector with two elements, 5 and 1. So I'd be totally OK if CPP2 just shadowed that constructor and only transpiled into the initializer list one. One less thing to teach. The ability to construct a vector of repeated elements can be exposed more clearly with a named constructor (if so desired), like `v := vector.repeated(5, 1)`.

**Comment by switch-blade-stuff** [comment]
> perhaps not even for aggregates

That being said, some amount of aggregate support is required, otherwise the classic C-style arrays, `std::array`, and other types that rely on aggregate initialization would not work outright (dont forget that `std::array` has *no* constructors).

From my point of view, `()` should be used for calling constructors, while `{}` should be used for aggregates & lists, but never constructors.

**Comment by JohelEGP** [comment]
That's fine, too. Though I recall @hsutter presenting about generalizing `()`.

**Comment by switch-blade-stuff** [comment]
> The ability to construct a vector of repeated elements can be exposed more clearly with a named constructor (if so desired), like v := vector.repeated(5, 1).

The issue I can see with this approach is that it would require custom treatment of container types (vectors, strings, sets, etc.). This means that if a third-party library provides a custom container with a similar "size & value" constructor, the patchfix wouldn't work and we will still have the same old issue.

And generalizing the size & value constructor through concepts does not seem like a very reliable approach either.

**Comment by switch-blade-stuff** [comment]
@JohelEGP I would prefer to use `()` for aggregates too (except for init-lists), but that will definitely break arrays unfortunately, since those rely on aggregates and not init-lists.

Imagine this:
```
v : vector = {1, 2, 3}; // vector of 1, 2, 3
a1 : array = {1, 2, 3}; // error
a2 : array = (1, 2, 3); // array of 1, 2, 3
```

That would be pretty confusing, but would make sense from the technical perspective, since `std::vector` is being initialized via an init-list, while `std::array` is being initialized via as an aggregate.

**Comment by jcanizales** [comment]
>> The ability to construct a vector of repeated elements can be exposed more clearly with a named constructor (if so desired), like v := vector.repeated(5, 1).

> The issue I can see with this approach is that it would require custom treatment of container types

No that proposal wouldn't require any custom treatment. The proposal would be that the syntax `v: T = (a, b)` transpiles to `T v{a, b}` always. So the problematic constructor of `std::vector` is shadowed and that might be a plus instead of a loss. It works the same for `std::vector` as for any third party container: if a third party offered the same bad API as `std::vector`, they get the same fix.

Writing a member function `v := vector.repeated(5, 1)` is one possibility of many if one wants to offer that constructor, but it's not up to the language to do it. Doesn't hurt to offer it, though.

**Comment by switch-blade-stuff** [comment]
> Writing a member function v := vector.repeated(5, 1) is one possibility of many if one wants to offer that constructor, but it's not up to the language to do it. Doesn't hurt to offer it, though.

I suppose that would work, yes. Although I still would prefer the `()` for constructors and `{}` for lists proposal since then you have clear visual separation of what is an element list and what is a constructor.

**Comment by hsutter** [comment]
Well, I wasn't planning to work on this today, but now you've all got me thinking about and I had intended to at some point think again about whether to address the `vector<NumericType>` two-parameter constructor issue... :)

So I've been paging back in my earlier thinking about this, reading your suggestions, and trying out two alternate implementations.

Here are some interim incomplete thoughts:

### Option 0: Status quo, and you just can't call the weird two-parameter non-list constructor on `vector<NumericType>`.

AFAIK problems like that constructor presents are super rare, and a "bad API" design smell anyway. We already teach people not to write functions that are overloaded on initializer lists and non-initializer lists in this ambiguous way, especially constructors. In the standard library I'm only aware of it arising for that one small set of `vector<NumericType>` specializations, and then for historical reasons -- that constructor was added before initializer lists existed.

For those reasons, I was delaying bothering to implement anything different, and was waiting to see if more use cases would crop up... I'd rather avoid complicating the language for the sake of a very rare and already-discouraged case, and if future experience showed that no other problems  arose, then "for the sake of `vector<int>`" felt like the tail wagging the dog and I'd be inclined not to bother.

One other case that did come up is call sites like `call_something( x, y, {1, 2, 3} )` that want to pass a list as an argument, where using parens instead like `call_something( x, y, (1, 2, 3) )` could create visual confusion especially for one-element lists (could look like a redundantly parenthesized expression). If there's an argument for having some way to explicitly write "this argument is a list of things," this use case seems like the stronger one to design for, and it maybe leans in favor of `{` `}`.

### Option 1: Use double parens.

As I mentioned above, I'd been noodling on that. But one of the reasons I didn't implement it yet because I was also concerned about the parens-as-precedence potential visual confusion, and possibly others. Elegant in one way, but could be overburdening parens in other places.

### Option 2: Use `{` `}` for lists.

Braces are less overburdened than parentheses. As @SebastianTroy noticed above, so far I've been reserving braces for scopes only. But using braces also for lists doesn't seem like it would conflict with scopes -- I can't offhand think of where they'd overlap in the grammar, should be fine even for expression-statements (I did have to write back-out logic for expression-statement and statement to make it work). And it has benefits: it's arguably clearer to pass anonymous lists as argument (primarily), it gives a bit of familiarity with today's C++ in a place where today's C++ is fine (secondarily)... and, yes, it lets people invoke `vector<int>`'s weird constructor (distant last place).

---

I've spent some time today implementing both (1) and (2) just to try them out and see their impact on the regression tests.
- (1) requires changing `(` `)` to `((` `))` in a bunch of tests. 
- (2) requires changing `(` `)` to `{` `}` in a bunch of tests (or `({` `})`, I tried that first).
- Both let call sites that want to pass a list as an argument spell it as `call_something( x, y, {1, 2, 3} )` which might be better than `call_something( x, y, (1, 2, 3) )` especially for single-element lists.

On the whole, I've learned some more from trying out (1) and (2), but this exercise has given me renewed appreciation for (0). :)

Anyway, this is a very-interim weak brain dump, and still in progress... I could change my mind about it a few minutes from now. But since this thread is continuing I thought I'd share a snapshot of thinking...

**Comment by SebastianTroy** [comment]
Could we use square [ ] brackets for lists/aggregates instead? Does that lead to ambiguous/difficult code with operator[]? The same symbols would be used to both define a list and access it, so

item : int = [0, 2, 44, 1][2]; // Item's value would be 44

I believe there are plans to deprecate comma operators inside operator[], so multidimensional array access would look like

item2 : int = [[8, 4], [5, 2]][0, 1]; // Item2's value would be 4

I do like cpp2's idea of removing curly brace initialisation, and focusing on uniform initialisation with just round braces, there are too many ways to initialise values in cpp1.

I do agree that the simplest answer was as Herb suggested, add more braces to denote the difference between an aggregate and a single value parameter, but also agree that this can lead to instances with a lot of consecutive opening or closing round (((  ))) braces, which gets hard to read.

v1 : vector<int>(5, 1); // [1, 1, 1, 1, 1]
v2 : vector<int>((5, 1)); // [5, 1]

It gets the job done, there's no new or special syntax, there's a lot of braces with multi dimensional arrays, but there always was anyway

v3 : vector<int>([5, 1]); // as proposed at the top

Doesn't really add all that much tbh, but it does make the outermost braces more obvious for messier examples

vector<vector<int>>([[5, 1], [5, 1], [5, 1]]);
vector<vector<int>>(((5, 1), (5, 1), (5, 1)));

I think its neat in that [x, y, z] must be a list or a list access, and that (x, y, z) must either be a function signature or a function call.

Thoughts?




On 2 January 2023 22:33:09 SwitchBlade ***@***.***> wrote:

@JohelEGP<https://github.com/JohelEGP> I would prefer to use () for aggregates too (except for init-lists), but that will definitely break arrays unfortunately, since those rely on aggregates and not init-lists.

Imagine this:

v : vector = {1, 2, 3}; // vector of 1, 2, 3
a1 : array = {1, 2, 3}; // error
a2 : array = (1, 2, 3); // array of 1, 2, 3


That would be pretty confusing, but would make sense from the technical perspective, since std::vector is being initialized via an init-list, while std::array is being initialized via as an aggregate.

**Comment by jcanizales** [comment]
Per @wolfseifert 's experiment, the current status quo would have to be changed no matter what:

```c++
main: () -> int = {
  a: std::vector<int> = (5, 1); // vector of five and one
  b := std::vector<int>(5, 1); // vector of five ones
}
```
currently transpiles to:
```c++
std::vector<int> a { 5, 1 };  // vector of five and one
auto b { std::vector<int>(5, 1) }; // vector of five ones
```
(Confusing!)

For **Option 0**, the second case would need to be changed to produce:
```c++
auto b { std::vector<int>{5, 1} };
```

**Comment by jcanizales** [comment]
Related consideration: Do you already have a plan in mind for the syntax of designated initializers, @hsutter ? That might be what "forces" CPP2 to use `{ }` for lists and aggregates anyway.

**Comment by switch-blade-stuff** [comment]
@SebastianTroy hmmm.. that tastes very... python-y

But to me the brackets proposal seems to hit the same issue as the double-parentheses one, in that the brackets are commonly associated with indexing, and using them for list initialization could create confusion.

**Comment by switch-blade-stuff** [comment]
@hsutter
> But using braces also for lists doesn't seem like it would conflict with scopes -- I can't offhand think of where they'd overlap in the grammar

I dont think there is any overlap, no, since scopes do not appear anywhere a prvalue might. Also, using braces for lists will have a good mental parallel to scopes: a scope is a collection of statements, while a list/aggregate is a collection of expressions.

**Comment by msadeqhe** [comment]
I've read [this issue in Carbon language](https://github.com/carbon-language/carbon-lang/issues/1462#issuecomment-1190662094) about conflicting `if/for/...` statements with `{}` if parenthesis are not mandatory in `if/for/...` statements.

**Comment by charlieman** [comment]
```
v1: vector<int> = 5, 1;   // size == 5, same as "= ((1, 1, 1, 1, 1))"
v2: vector<int> = _, (5, 1); // size == 2
```

**Comment by hsutter** [comment]
Thanks, everyone, for all the comments!

@SebastianTroy 
> Could we use square [ ] brackets for lists/aggregates instead?

Let's add that option to the list, and I've now implemented (3) too to try it out:

### Option 3: Use `[` `]` for lists.

This syntax more visually connotes a list-of-things, it's symmetric with `[` `]` for element access of lists/arrays, and it's easier to see the difference between `(` `)` and `[` `]` (at least for me) than between `(` `)` and `{` `}`.

Its biggest drawback is one I'd encountered with `[` `]` before (*), that doing this creates some extra work to distinguish between `[` `]` used for subscripting (which have to be preserved as `[` `]` when lowering to Cpp1) and `[` `]` used for lists (which in this option would need to be emitted as `{` `}` to Cpp1). My learnings from such situations so far has generally been that "when it's harder for cppfront to know whether something is A or B, it's generally also harder for the programmer to see whether it's A or B" plus "and usually hits the 'have to do name lookup to figure out what it means' problem" (a key issue Cpp2 is aiming to solve is that name lookup is never needed for parsing, but it also shouldn't be needed to understand what an expression means). This highlighted to me that this path would make `thing[1]` at least visually ambiguous... if `thing` is an object with a subscript operator then it's a subscript operator call, but if `thing` is a type then it can be a function-style cast (see below).

(*) A much worse case of this is one of the reasons I went away from using `[` `]` for template argument lists, which I had been using in earlier drafts of Cpp2... I didn't want to have to do name lookup to parse an expression like `thing[1]`. This would be less bad than that, as at least both meanings of `thing[1]` would still be an expression, but it would still be a visual ambiguity that the human and the compiler would need to resolve each time.

### Current resolution: (0) Status quo

I've implemented all three of (1) (2) and (3) and tried them out in the regression tests, and it's been an interesting experiment. After writing a code in each of the other three syntaxes, and even kind of liking each of those syntaxes, I found that when I switched back to status quo it felt even nicer still to just have `(` `)` and not have to teach or think about `((` `))`, `{` `}`, or `[` `]` (each of which was at least a second thing to teach, and most of which additionally caused some secondary effects or at least visual ambiguities). Maybe it's just me, but after trying the others, "just parens" surprisingly felt like a small breath of fresh air. It felt good.

I'm still going to be open to further experience, but for now I don't think it's worth adding a language feature primarily to make `vector<NumericType>`'s disparaged two-parameter constructor work as a first-class initializer (you can still get it using a function-style expression, because Cpp1 and current Cpp2 allow using a type name as a callable name in the postfix-expression syntax).

The more potentially compelling use case for `{` `}` expression-lists in the future is the one I mentioned above -- it could be useful to have an explicit initializer-list expression syntax, which would allow calls like "f( {1, 2, 3} )`. I've kept the code changes that would enable that to make it easier to try that in the future(not now), but I've deliberately disabled accepting the `{` `}` there for the time being.

---

> Do you already have a plan in mind for the syntax of designated initializers, @hsutter ? That might be what "forces" CPP2 to use { } for lists and aggregates anyway.

I've been deprioritizing them because they're not addressing one of the pressing problems with C++ and so can be looked at later. When I do look at them, I plan to consider them together with named arguments as a general feature, and avoid doing something special for constructors that doesn't work the same way for other function calls.

Thanks everyone!

**Comment by wolfseifert** [comment]
So the only way to pass a <code>vector</code> of <code>5</code> and <code>1</code> to a function <code>print</code> is like this:
<pre>
a: std::vector<int> = (5, 1);
print(a); // prints 5 1
</pre>
Neither
<pre>
print(std::vector<int>(5, 1));  // prints 1 1 1 1 1
</pre>
nor
<pre>
print(: std::vector<int> = (5, 1));  // does not transpile
</pre>
do the trick. Or did I miss something?

**Comment by jcanizales** [comment]
@wolfseifert does `print((5, 1))` work?

**Comment by wolfseifert** [comment]
No, it does not. I will open a new issue for this later.

**Comment by hsutter** [comment]
>     print(: std::vector = (5, 1));  // does not transpile

I've been meaning to allow unnamed objects, and seeing you actually trying to write the code pushed me over the edge to implement it. :)

This now works:

```
print: (x:_) = {
    for x do :(elem:_) = {
        std::cout << elem << " ";
    }
}

main: ()->int = {
    print( : std::vector = (5,1) );   // prints "5 1"
}
```

**Comment by msadeqhe** [comment]
Is it possible to infer type from function declaration?
```
print :(x :std::vector<int>) = {
    for x do :(elem:_) = {
        std::cout << elem << " ";
    }
}

main :() ->int = {
    print(:= (5,1));   // prints "5 1"
}
```

**Comment by wolfseifert** [comment]
How I understand this so far: 
 

- <pre>x: std::vector<int> = (5, 1);</pre> is the current "standard" way of initializing a <code>vector</code>. It is transpiled to <pre>std::vector<int> x {5, 1};</pre>
- <pre>y: std::vector<int> = : std::vector = (5, 1);</pre> is a more cumbersome way of achieving the same thing. It is transpiled to <pre>std::vector<int> y {std::vector{5, 1}};</pre> This variant however makes a lot of sense for something like <pre>y: std::vector<int> = : myvector = (5, 1);</pre> where <code>myvector</code> is a subclass of <code>vector</code> but <code>y</code> must be of type <code>vector</code>.
- <pre>z: std::vector<int> = := (5, 1);</pre> is transpiled to <pre>std::vector<int> z {auto{5, 1}}</pre> which does not C++ compile.

**Comment by filipsajdak** [comment]
What do you think about that:

```cpp
print( (5,1) as std::vector ); // for generic functions; prints "5 1"

taking_vector( forward (5,1) ); // assuming that this function expects a vector and 
                                // that we can deduce the type from the function declaration
```

For the `std::vector` "bad API" issue, maybe we should think about correcting it on the cppfront side e.g., like that:
```cpp
main: () -> int = {
    print( std::vector<int>().create(5, 1) ); // prints 1 1 1 1 1
    // or ...
    print( create<std::vector>(5, 1) ); // prints 1 1 1 1 1
}

template <typename X>
auto create(std::vector<X>&&, std::size_t size, X&& value = X{}) -> decltype(auto)
{
    return std::vector<X>(size, value);
}

template <template <typename...> class V, typename X>
auto create(std::size_t size, X&& value = X{}) -> decltype(auto)
{
    return V<X>(size, value);
}

print: (x:_) = {
    for x do :(elem :_) = {
        std::cout << elem << " ";
    }
    std::cout << std::endl;
}

#include <vector>
```

Names can be adjusted (`resize` cannot be used instead of `create` as `vector` already has a `resize` method, and UFCS prefers methods over free functions).

**Comment by filipsajdak** [comment]
I have just realized that syntax for unnamed objects
```cpp
: std::vector = (5, 1)
```
Follows the same syntax as unnamed functions (aka lambdas):
```cpp
: (i:int) -> int = { return i+1; }
```

**Comment by hsutter** [comment]
Re syntax: Yes, in both cases it's the identical syntax as a named object/function, but without the name.

I think we've bottomed out on this for now. Thanks, everyone!

**Comment by JohelEGP** [comment]
<https://github.com/hsutter/cppfront/issues/284#issuecomment-1553081710> means it's possible to call `std::vector<int>`'s `(int, int)` constructor: <https://cpp2.godbolt.org/z/Pvj8fj1he>.
```Cpp2
using std::vector;
main: () = {
  z := 0.vector(0);
}
```

**Comment by AbhinavK00** [comment]
That's a nice hack.
But the fact that it took this long to come up with a solution shows we need some kind of change.
